lambdaclass/mirra_backend
# Refactor CampaignProgress to SuperCampaignProgress

Frontend PR: https://github.com/lambdaclass/champions_of_mirra/pull/314

Closes https://github.com/lambdaclass/champions_of_mirra/issues/193

Refactors CampaignProgress to SuperCampaignProgress (more details in the issue).

Moves level advance logic from `GameBackend.Users` to `Champions.Campaigns`, since we were doing a lot of champions-specific operations there.

# To Test

Create User
`{:ok, user} = Champions.Users.register("User")`
Add units
`team1 = Enum.map(1..6, fn slot -> GameBackend.Units.insert_unit(%{character_id: GameBackend.Units.Characters.get_character_by_name("Muflus").id, level: 99999, tier: 1, rank: 1, selected: true, user_id: user.id, slot: slot}) |> elem(1) |> GameBackend.Repo.preload(character: [:basic_skill, :ultimate_skill]) end)`
Get campaigns
`{:ok, [campaign_1, campaign_2 | _]} = Champions.Campaigns.get_campaigns`
Get levels
`[level1, level2, level3, level4, level5, level6, level7, level8, level9, level10] = campaign_1.levels`
For each level, fight it, win it and check that the CampaignProgress for user has changed to the next level.
`Champions.Battle.fight_level(user.id, level1.id)`
`{:ok, scp} = GameBackend.Campaigns.get_super_campaign_progress user.id, campaign_1.super_campaign_id`
`scp.level_id == level2.id` should be true
When you beat level 10, the last of the campaign, CampaignProgress should change level and campaign
